Faster singling out attack (20x-30x) via polars

AwesomeLemon opened the pull request:
The singling out attack implementation is slow because queries are constructed and evaluated sequentially, one-by-one. 

In this pull request, queries are instead evaluated in batches, in parallel. This is enabled by [polars](https://pola.rs/), an efficient pandas alternative. The SinglingOutEvaluator still accepts pandas dataframes but they are converted to polars dataframes that are used to construct and evaluate queries.

Time was measured with the code in ``notebooks/measure_time_singling_out.ipynb``: 2000 queries are constructed for the Adult dataset using either 3 or 12 auxiliary columns (the minimum and the maximum number of auxiliary columns used in the practical attack in the Anonymeter paper). Query generation and evaluation is repeated 3 times, mean and standard deviation are reported both for the current pandas implementation and the proposed polars implementation. The Standard_E4d_v4 Azure machine (4 CPU cores) was used (using a CPU with more cores is expected to lead to even larger speed-ups).

The results:

| Number of columns | Pandas (mean ± std) | Polars (mean ± std) | Speed-up |
|-------------------|---------------------|---------------------|----------------------------|
| 3                 | 527.89 ± 1.34 s     | 17.46 ± 0.03 s      | 30.2×                     |
| 12                | 61.20 ± 0.03 s      | 3.01 ± 0.10 s       | 20.3×                     |

Additionally, ``seed`` parameter has been added to the SinglingOutEvaluator to enable reproduciblity, and a bug has been fixed where in the case that fewer queries were successfully constructed than expected, the expected rather than actual number of queries was used in the risk estimation, which led to underestimation.

To ensure that the code functions correctly (in addition to tests passing), the old and the new implementations were used to estimate the risks of the synthetic data included in the repository, and artificially constructed synthetic data with 20% leakage. The code in ``notebooks/singling_out.ipynb`` was used, with each experiment repeated 5 times. The results are similar between the implementations.

|Data | Pandas | Polars |
|---|---|---|
|CTGAN| ![image](https://github.com/user-attachments/assets/f565303f-703b-4a40-9944-e69c25747813) | ![image](https://github.com/user-attachments/assets/a545e54f-37ab-4c84-b3d6-c53c02c5dec4) |
|Artificial| ![image](https://github.com/user-attachments/assets/ad949632-5da5-4825-b3c2-c6b07187eea1) | ![image](https://github.com/user-attachments/assets/387e883c-0608-40a5-acb7-76e384b122af) |

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/statice/anonymeter?pullRequest=42) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/statice/anonymeter?pullRequest=42) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/statice/anonymeter?pullRequest=42) it.</sub>

MatteoGiomi commented on src/anonymeter/evaluators/singling_out_evaluator.py:
```diff
@@ -2,129 +2,168 @@
-    else:
-        query = f"{col} == {val}"
+def _query_from_record(
+    record: dict,  # or pl.Series, but dict is often easier for single rows
```
If `record` can be either `dict` or `pl.Series`, why not typing it accordingly?

```suggestion
    record: Union[dict, pl.Series],  # dict is often easier for single rows
```

MatteoGiomi commented on src/anonymeter/evaluators/singling_out_evaluator.py:
```diff
@@ -2,129 +2,168 @@
+
+            expr_col = op(pl.col(col), val)
+
+        elif dtypes[col] == pl.Categorical:
+            expr_col = pl.col(col) == val
```
This case could be grouped with the boolean one since they result in the same `expr_col`.

MatteoGiomi commented on src/anonymeter/evaluators/singling_out_evaluator.py:
```diff
@@ -2,129 +2,168 @@
+        exprs.append(e)
 
-    return " & ".join(query)
+    return reduce(operator.and_, exprs)
```
```suggestion
   # use bitwise and
    return reduce(operator.and_, exprs)
```

if I understand correctly.

MatteoGiomi commented on src/anonymeter/evaluators/singling_out_evaluator.py:
```diff
@@ -2,129 +2,168 @@
+        ).tolist()
+
+        queries.append(
+            _random_query(unique_values, selected_cols, column_types)
```
I think you forgot to pass the rng around here (see my other comments about avoiding defaults in low-level funcitons). Also, I think named arguments are much safer in general.

```suggestion
            _random_query(
                unique_values=unique_values,
                cols=selected_cols,
                column_types=column_types,
                rng=rng_to_use)
            
```

MatteoGiomi commented on src/anonymeter/evaluators/singling_out_evaluator.py:
```diff
@@ -209,54 +260,62 @@ def fit_correction_term(df: pd.DataFrame, queries: List[str]) -> Callable:
         """
-
-        if query not in self._set:
-            counts = safe_query_counts(query=query, df=df)
```
I still have to see how this is used in the code, but I liked the idea of having the "safe" counts done internally and not left to the user responsibility...

MatteoGiomi commented on src/anonymeter/evaluators/singling_out_evaluator.py:
```diff
@@ -209,54 +260,62 @@ def fit_correction_term(df: pd.DataFrame, queries: List[str]) -> Callable:
 
-def univariate_singling_out_queries(df: pd.DataFrame, n_queries: int) -> List[str]:
+def univariate_singling_out_queries(
+    df: pl.DataFrame, n_queries: int, rng: Optional[np.random.Generator] = None
```
Add docstring for `rng`

MatteoGiomi commented on src/anonymeter/evaluators/singling_out_evaluator.py:
```diff
@@ -330,66 +402,131 @@ def multivariate_singling_out_queries(
+            for _ in range(batch_size)
+        ]
+
+        queries_batch = [
+            _query_from_record(
+                record,
+                dtypes_dict,
+                columns,
+                medians_dict,
+                rng,
+            )
+            for record, columns in zip(records, selected_columns)
+        ]
```
NIT:

```suggestion
        queries_batch = [
            _query_from_record(
                records_batch,
                dtypes_dict,
                columns,
                medians_dict,
                rng,
            )
            for records_batch, columns in zip(records, selected_columns)
        ]
```

and consider naming the arguments.

MatteoGiomi reviewed (commented):
Thanks a lot for this contributions, indeed the speed of the SO evaluator was the main pain point of Anonymeter. Also thanks for checking with experiments that everything works as before.

 It's a lot of nice code and I've left some comments, none of them really major. There were some disconnected random number generator which might explain some of the small differences in your results. 


When I try to look at the notebooks I get: "This commit does not belong to any branch on this repository, and may belong to a fork outside of the repository.".

AwesomeLemon commented on src/anonymeter/evaluators/singling_out_evaluator.py:
```diff
@@ -268,44 +327,57 @@ def univariate_singling_out_queries(df: pd.DataFrame, n_queries: int) -> List[st
+                        pl.col(col) <= col_min,
+                        pl.col(col) >= col_max,
+                    ]
+                )
```
You were correct to question this code. I've looked into it, and apparently if the entire column is null, polars throws an error when min/max is called. So I've implemented the logic closer to the original code: the queries comparing to min/max are only added if the column has more than 0 non-null values.

AwesomeLemon replied:
Of course, I've now added a test for a column full of null values, and a test for a mixed column with null and non-null values

AwesomeLemon commented on src/anonymeter/evaluators/singling_out_evaluator.py:
```diff
@@ -330,66 +402,131 @@ def multivariate_singling_out_queries(
+        )
+
+        # Extract all records in bulk
+        records = df[random_indices].to_dicts()
```
In polars, there is no index and no ``.iloc`` (see [here](https://docs.pola.rs/user-guide/migration/pandas/) for details). The code I wrote should be equivalent to using ``.iloc`` in pandas, so I did not change anything here

AwesomeLemon commented:
Thank you for taking the time to give feedback on my code! 

I've implemented most of it, here are the key changes:

1. UniqueSinglingOutQueries modified to evaluate queries inside.
2. Fixed an issue where an all-null column would throw an error in univariate query construction due to min/max calls.
3. Random generator is specified in one place and passed to the functions that need it.

I've resolved the comments with your suggestions that I've straightforwardly implemented. I've answered to the comments where the changes I made were not completely trivial. 

The only strong point of disagreement is about dividing by the required/actual number of attacks - I've jotted down my thoughts as a response to your comment there.

I'm a bit confused why you can't see the notebooks: I see them both in this pull request, and in my fork (see [singling_out.py](https://github.com/AwesomeLemon/anonymeter/blob/7b6e6937a5e7e8a6c7127887ba702c080730e0eb/notebooks/singling_out.ipynb) and [measure_time_singling_out.ipynb](https://github.com/AwesomeLemon/anonymeter/blob/7b6e6937a5e7e8a6c7127887ba702c080730e0eb/notebooks/measure_time_singling_out.ipynb))

MatteoGiomi commented on src/anonymeter/evaluators/singling_out_evaluator.py:
```diff
@@ -502,34 +649,58 @@ def evaluate(self, mode: str = "multivariate") -> "SinglingOutEvaluator":
-        self._queries = _evaluate_queries(df=self._ori, queries=queries)
         self._n_success = len(self._queries)
 
+        # if you can't generate enough, warn the user and proceed with what you have - should divide by the attacks you have, not what you wish you had!
```
I know what you mean, we spend quite some time thinking of what's the best normalization to use in these cases. Still, I think that normalizing by the requested number of queries (instead of by the ones found by the attacker) is preferable.

Think of why this happens: the attacker fails to fulfill their task (generating all the requested queries) because the chosen algorithm and parameters (number of columns) in combination with the specific synthetic data did not allow them to do so. It's a failure of the attacker and hence it is ok if it results in a "lower" risk estimates. 

If the synthetic data is harder to attack with the chosen algorithm/parameters, the solution is to try harder (increase `max_attemps`) and/or change hyper parameters (number of columns), rather than the normalization.

I am open to discuss this further, but maybe we can move this change to another PR as not to block this one.

MatteoGiomi commented:
Hi @AwesomeLemon, thanks again for your code. While sleeping over it, two questions came to my mind: 
- how (if possible) can the user control the amount of paralelization introduced by `polars`?
-

MatteoGiomi commented on src/anonymeter/evaluators/singling_out_evaluator.py:
```diff
@@ -209,54 +264,67 @@ def fit_correction_term(df: pd.DataFrame, queries: List[str]) -> Callable:
     """
-    sizes, successes = _measure_queries_success(df=df, queries=queries, n_repeat=5, n_meas=10)
+    sizes, successes = _measure_queries_success(
+        df=df, queries=queries, n_repeat=5, n_meas=10
```
now that the evaluation of the queries is much faster, we can improve the quality of the results by increasing a bit these parameters to make the fit more robust.. With a ~20x speedup we could maybe even go to `n_repeat=10` and `n_meas=20`, wdyt?

MatteoGiomi replied:
Yes, good point.

MatteoGiomi commented:
one question about `polars`. If it possible (and if so how) to control the amount of parallelization? I might have missed it in the code..

AwesomeLemon commented:
> one question about `polars`. Is it possible (and if so how) to control the amount of parallelization? I might have missed it in the code..

By default, polars uses all the cores. To reduce the number of used cores, it seems that the ``POLARS_MAX_THREADS`` environment variable has to be exported (https://github.com/pola-rs/polars/issues/1718)

MatteoGiomi reviewed (approved):
Thank you very much @AwesomeLemon for this very important contribution and for adressing all the comments in the PR.. it's a really solid piece of work and I'm very happy to approve this PR. Feel free to merge!
